src/plugins/RepairMan.cpp

```diff
@@ -53,3 +95,18 @@ void RepairMan::OnUnitDestroyed(const sc2::Unit* unit_, Builder* builder_) {
             return;
     }
 }
+
+void RepairMan::AssignRepairTask(const sc2::Unit& damaged_) {
+    std::function<bool(const Worker&)> moreThanHalfHealth =
+    [](const Worker& worker) {
+        const sc2::Unit* unit = gAPI->observer().GetUnit(worker.Tag());
+        return unit ? unit->health > unit->health_max * 0.5f : false;
+    };
+
+    const sc2::Point2D repair_location(damaged_.pos);
+    Worker* worker = gHub->GetClosestFreeWorker(repair_location, moreThanHalfHealth);
```

Comment:
If such filtering really needed, I suggest to do it in a more functional way, i.e. get list of workers with required health value. Then filter it by distance.

Reply:
Do you mean:
Get const ref of Cache<Worker> m_free_workers (Have to add Hub::GetFreeWorkers() for this).
Iterate through m_free_workers and put Filtered workers in a temp Cache<Worker>.
Then use Cache::GetClosestTo on the filtered Cache

I believed you would want me to *avoid* creating copies of the Workers. 

If I instead filter to a list of Worker*, I can't use the Cache::GetClosestTo and have to do a manual distance check. Is this what you are suggesting?

To me, passing in a filter is the simple and flexible way to achieve the goal. As it's passing a function as a first-order object, would it not be considered the more "functional way"?

Reply:
First of all, I don't like the current implementation of GetClosestFreeWorker as it always checks do we have functor or not. It means that technically we have two different methods wrapped inside one. The functional way allows to solve this task in a more elegant way, i.e.

a. Get free workers list.
b. Create new list of healthy workers.
c. Find the closest one.

As a result at least one copy of units array will be created, but it'll not provide large overhead I think. Also, we can mitigate it by using std::move, although I can't give particular advice on it as sadly I am not familiar with this new mechanic.

Another simple improvement could be inclusion of pointer to sc2::Unit inside GameObject class. The original design assumed that usually we don't need the pointer but can use the current task most of the time. Now it is not enough so we can avoid GetUnit by tag call with this small fix. The API guaranties that the pointer will be kept alive whole game.

Reply:
Originally, I did create GetClosetFreeWorker / GetClosestTo Filtered versions as separate additional functions, but then I thought you would not like as it is the exact same code with just 2 extra checks, so I combined them.

I'll make filtered copy array and add Unit*.

Reply:
I see the GameObject has two ToUnit functions which return a dummy unit with only a valid Tag, for sending to API. I'm suggesting to rename these to 'ToDummyUnit' so they are not confused with a Unit/Unit* to a real unit with full data. (I was confused and had to figure that out earlier)

Reply:
Please, don't do it. The only reason `ToUnit` exists is the API which doesn't support command execution by tag. That should be fixed on the API's side and `ToUnit` should be removed.
The simple solution: make the unit pointer public without additional getters.

Reply:
@alkurbatov  I looked up std::move, and it is for used to move a temporary object (rvalue reference T&&) into a permanent variable without copy-constructor, but it sets the old object to null, so if used on m_free_workers, it would zero out the object.

"
1) In C++11, in addition to copy constructors, objects can have move constructors.
(And in addition to copy assignment operators, they have move assignment operators.)
2) The move constructor is used instead of the copy constructor, if the object has type "rvalue-reference" (Type &&).
3) std::move() is a cast that produces an rvalue-reference to an object, to enable moving from it.

It's a new C++ way to avoid copies. For example, using a move constructor, a std::vector could just copy its internal pointer to data to the new object, leaving the moved object in a moved-from state, therefore not copying all the data. This would be C++-valid.
https://en.wikipedia.org/wiki/C%2B%2B11#Rvalue_references_and_move_constructors
"
https://stackoverflow.com/questions/3413470/what-is-stdmove-and-when-should-it-be-used

Reply:
You need to create a new object after filtration of m_free_workers i.e. a vector with > 50% hp workers. This tmp object could be returned to caller with std::move. It allows to avoid doing 1 on 2 copies.